src/stan/services/sample/hmc_nuts_dense_e_adapt.hpp

```diff
@@ -156,6 +155,219 @@ int hmc_nuts_dense_e_adapt(
+        stepsize, stepsize_jitter, max_depth, delta, gamma, kappa, t0,
+        init_buffer, term_buffer, window, interrupt, logger, init_writer[0],
+        sample_writer[0], diagnostic_writer[0]);
+  } else {
+    using sample_t = stan::mcmc::adapt_dense_e_nuts<Model, boost::ecuyer1988>;
+    std::vector<boost::ecuyer1988> rngs;
+    rngs.reserve(num_chains);
+    std::vector<std::vector<double>> cont_vectors;
+    cont_vectors.reserve(num_chains);
+    std::vector<sample_t> samplers;
+    samplers.reserve(num_chains);
+    try {
+      for (int i = 0; i < num_chains; ++i) {
+        rngs.emplace_back(util::create_rng(random_seed, init_chain_id + i));
+        cont_vectors.emplace_back(util::initialize(model, *init[i], rngs[i],
+                                                   init_radius, true, logger,
+                                                   init_writer[i]));
+        Eigen::MatrixXd inv_metric = util::read_dense_inv_metric(
+            *init_inv_metric[i], model.num_params_r(), logger);
+        util::validate_dense_inv_metric(inv_metric, logger);
+
+        samplers.emplace_back(model, rngs[i]);
+        samplers[i].set_metric(inv_metric);
+        samplers[i].set_nominal_stepsize(stepsize);
+        samplers[i].set_stepsize_jitter(stepsize_jitter);
+        samplers[i].set_max_depth(max_depth);
+
+        samplers[i].get_stepsize_adaptation().set_mu(log(10 * stepsize));
+        samplers[i].get_stepsize_adaptation().set_delta(delta);
+        samplers[i].get_stepsize_adaptation().set_gamma(gamma);
+        samplers[i].get_stepsize_adaptation().set_kappa(kappa);
+        samplers[i].get_stepsize_adaptation().set_t0(t0);
+        samplers[i].set_window_params(num_warmup, init_buffer, term_buffer,
+                                      window, logger);
+      }
+    } catch (const std::domain_error& e) {
+      return error_codes::CONFIG;
+    }
+#ifdef STAN_THREADS
+    tbb::parallel_for(
+        tbb::blocked_range<size_t>(0, num_chains, 1),
+        [num_warmup, num_samples, num_thin, refresh, save_warmup, num_chains,
+         &samplers, &model, &rngs, &interrupt, &logger, &sample_writer,
+         &cont_vectors,
+         &diagnostic_writer](const tbb::blocked_range<size_t>& r) {
+          for (size_t i = r.begin(); i != r.end(); ++i) {
+            util::run_adaptive_sampler(
+                samplers[i], model, cont_vectors[i], num_warmup, num_samples,
+                num_thin, refresh, save_warmup, rngs[i], interrupt, logger,
+                sample_writer[i], diagnostic_writer[i], i + 1, num_chains);
+          }
+        },
+        tbb::simple_partitioner());
+#else
+  for (size_t i = 0; i != num_chains; ++i) {
+    util::run_adaptive_sampler(
+        samplers[i], model, cont_vectors[i], num_warmup, num_samples,
+        num_thin, refresh, save_warmup, rngs[i], interrupt, logger,
+        sample_writer[i], diagnostic_writer[i], i + 1, num_chains);
+  }
+#endif
```

Comment:
@wds15 I was looking at the code for `reduce_sum()` and was getting a bit confused. Do I need to have code like this for when `STAN_THREADS` is defined or not? Going through `tbb::parallel_for()` without `STAN_THREADS` caused the updated version of cmdstan (below) to segfault.

Reply:
Actually this type of code should not be necessary. What I think happens is that the TBB is spawning happily multiple threads, but it should not. So I would suggest to patch

https://github.com/stan-dev/math/blob/develop/stan/math/prim/core/init_threadpool_tbb.hpp

such that whenever `STAN_THREADS` is not set, then the TBB thread pool will be set to use only 1 thread. Then the above should be fine without the ifdef. Can you try that? That is patch the init function for the case `STAN_THREADS` is not set to always just use 1 thread and then disable the ifdef above.

Reply:
Yep this was the issue! Just put in a PR in math to fix this

Reply:
Cool. So we can get rid of the ifdef once this fix in math is in. Great.